keanu-project/src/main/java/io/improbable/keanu/vertices/dbltensor/TensorShapeValidation.java

```diff
@@ -0,0 +1,108 @@
+package io.improbable.keanu.vertices.dbltensor;
+
+import java.util.Arrays;
+import java.util.Set;
+
+import static java.util.stream.Collectors.toSet;
+
+public class TensorShapeValidation {
+
+    private TensorShapeValidation() {
+    }
+
+    /**
+     * This is a common function to check that tensors are either
+     * the same shape of the proposal in question OR scalar.
+     *
+     * @param proposalShape the tensor shape being validated
+     * @param tensors       the tensors being validated against
+     * @throws IllegalArgumentException if there is more than one non-scalar shape OR if the non-scalar shape does
+     *                                  not match the proposal shape.
+     */
+    public static void checkTensorsMatchNonScalarShapeOrAreScalar(int[] proposalShape, Tensor... tensors) {
+
+        Set<TensorShape> nonScalarShapes = getNonScalarShapes(tensors);
+
+        if (!nonScalarShapes.isEmpty()) {
+
+            boolean moreThanOneNonScalarShape = nonScalarShapes.size() > 1;
```

Comment:
the variable assignment here seems a bit over the top. I think it's fine to have:

```
if (nonScalarShapes.size() > 1) {
            throw new IllegalArgumentException("More than a single non-scalar shape");
}
```